### PR TITLE
Repo to query envelopes per day

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-1026

### Change description ###
EnvelopeSummary and EnvelopeSummaryRepository.

EnvelopeSummaryRepository will be called from the service layer which will be in charge of deriving 'from' and 'to' timestamps for a given date.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        NamedParameterJdbcTemplate jdbcTemplate,
+        EnvelopeSummaryMapper mapper
+    ) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.mapper = mapper;
+    }
+
+    public List<EnvelopeSummary> find(Instant from, Instant to) {
+        return jdbcTemplate.query(
+            "SELECT e.*, ev2.* "
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: is it not better to select specific columns we need? certainly help validating mapper

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+@Component
+public class EnvelopeSummaryMapper implements RowMapper<EnvelopeSummary> {
+
+    @Override
+    public EnvelopeSummary mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return new EnvelopeSummary(
+            rs.getString("container"),
+            rs.getString("file_name"),
+            rs.getTimestamp("file_created_at").toInstant(),
+            rs.getTimestamp("dispatched_at").toInstant(),
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java]: `dispatched_at` can be null. current statement will throw an error

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        // when
+        List<EnvelopeSummary> result = repo.find(
+            LocalDate.parse("2019-12-20").atStartOfDay().toInstant(UTC),
+            LocalDate.parse("2019-12-21").atStartOfDay().toInstant(UTC)
+        );
+
+        // then
+        assertThat(result.stream()
+                       .collect(toList()))
+            .extracting("container", "fileName", "fileCreatedAt", "dispatchedAt", "status", "lastEvent", "eventNotes")
```
Comment (doncem) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: i think we prefer lambda expression here - it's type/compile safe as well

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            .containsExactlyInAnyOrder(
+                tuple(CONTAINER_1, FILE_1_2, created2, dispatched, Status.DISPATCHED, DISPATCHED, null),
+                tuple(CONTAINER_1, FILE_1_3, created3, dispatched, Status.DISPATCHED, DELETED, null),
+                tuple(CONTAINER_2, FILE_2_1, created4, dispatched, Status.REJECTED, DUPLICATE_REJECTED, "Duplicate"),
+                tuple(CONTAINER_2, FILE_2_2, created5, dispatched, Status.REJECTED, DELETED_FROM_REJECTED, null),
```
Comment (doncem) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: rejected envelopes don't have dispatched timestamp

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java (hunk 1 above)
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: Maybe it would be best to list specific columns and give them aliases, as event and envelope have some overlap when it comes to column names.

Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: Also, we need to filter out envelopes from bulkscan container. This is probably the best place to do it.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EnvelopeSummary.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    private final String container;
+    private final String fileName;
+    private final Instant fileCreatedAt;
+    private final Instant dispatchedAt;
+    private final Status status;
+    private final Event lastEvent;
+    private final String eventNotes;
+
+    public EnvelopeSummary(
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EnvelopeSummary.java]: Perhaps last event creation date would be useful?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+            rs.getString("container"),
+            rs.getString("file_name"),
+            rs.getTimestamp("file_created_at").toInstant(),
+            rs.getTimestamp("file_dispatched_at") == null
+                ? null : rs.getTimestamp("file_dispatched_at").toInstant(),
+            Status.valueOf(rs.getString("status")),
+            rs.getBoolean("is_deleted"),
+            rs.getString("event") == null ? null : Event.valueOf(rs.getString("event")),
+            rs.getString("notes"),
+            rs.getTimestamp("event_created_at") == null
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java]: event creation is never null. notes though are. unless `getString` handles it there's nothing else to do :)

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                + "    FROM events "
+                + "    GROUP BY container, file_name "
+                + ") ev1 "
+                + "ON (e.container = ev1.container AND e.file_name = ev1.file_name) "
+                + "LEFT OUTER JOIN ("
+                + "    SELECT id, event, notes, created_at "
+                + "    FROM events"
+                + ") ev2 "
+                + "ON ev2.id = ev1.max_id "
+                + "WHERE e.container <> '" + BULKSCAN_CONTAINER_NAME + "' "
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: why not use prepared statement feature instead?

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                new EnvelopeSummary(CONTAINER_2,
+                                    FILE_2_3,
+                                    createdAt6,
+                                    dispatchedAt,
+                                    Status.DISPATCHED,
+                                    false,
+                                    null,
+                                    null,
+                                    null
+                )
```
Comment (doncem) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: first argument should be new line as well.

on other note: so many fields in assertion 🙊

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java (hunk 7 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: Have discussed this with Leszek, decided that we should keep using existing approach unless there are clear benefits in using a new approach

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java (hunk 6 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java]: only if there are events for the envelope, otherwise it is null

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java (hunk 7 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: Open for discussion, though. If there are good benefits, we might want to move to prepared statements from the approach we're currently using.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EventRecord.java [hunk 9]
```diff
@@ -26,4 +26,8 @@ public EventRecord(
         this.event = event;
         this.notes = notes;
     }
+
+    public long getId() {
+        return id;
+    }
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EventRecord.java]: Why do we need this getter?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EnvelopeSummary.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        String container,
+        String fileName,
+        Instant fileCreatedAt,
+        Instant dispatchedAt,
+        Status status,
+        boolean isDeleted,
+        Event lastEvent,
+        String eventNotes,
+        Instant eventCreatedAt
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EnvelopeSummary.java]: Since the type of last event is called `lastEvent`, maybe `eventNotes` and `eventCreatedAt` should also have `last` prefix?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java (hunk 6 above)
Comment (lgonczar) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java]: I think the envelope is created together with an event, in a single transaction. Not sure the mapper needs to be aware of it, though.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeRepository.java [hunk 11]
```diff
@@ -86,7 +86,10 @@ public UUID insert(NewEnvelope envelope) {
                 .addValue("fileName", envelope.fileName)
                 .addValue("fileCreatedAt", Timestamp.from(envelope.fileCreatedAt))
                 .addValue("status", envelope.status.name())
-                .addValue("dispatchedAt", Timestamp.from(envelope.dispatchedAt))
+                .addValue(
+                    "dispatchedAt",
+                    envelope.dispatchedAt == null ? null : Timestamp.from(envelope.dispatchedAt)
+                )
```
Comment (lgonczar) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeRepository.java]: We should have some tests around it. Perhaps it would be best to extract this bit into a separate PR then?

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+import static uk.gov.hmcts.reform.blobrouter.data.model.Event.DUPLICATE_REJECTED;
+import static uk.gov.hmcts.reform.blobrouter.data.model.Event.FILE_PROCESSING_STARTED;
+import static uk.gov.hmcts.reform.blobrouter.data.model.Event.REJECTED;
+
+@ActiveProfiles({"integration-test", "db-test"})
+@SpringBootTest
+public class EnvelopeSummaryRepositoryTest {
+    @Autowired private EnvelopeRepository envelopeRepository;
+    @Autowired private EventRecordRepository eventRecordRepository;
+    @Autowired private EnvelopeSummaryRepository repo;
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: Given two other repositories have long names, it would be good for this one to be called `envelopeSummaryRepository` for clarity.

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    private static final String CONTAINER_1 = "container1";
+    private static final String CONTAINER_2 = "container2";
+    private static final String BULKSCAN_CONTAINER = "bulkscan";
+    private static final String FILE_1_1 = "file_name_1_1";
+    private static final String FILE_1_2 = "file_name_1_2";
+    private static final String FILE_1_3 = "file_name_1_3";
+    private static final String FILE_2_1 = "file_name_2_1";
+    private static final String FILE_2_2 = "file_name_2_2";
+    private static final String FILE_2_3 = "file_name_2_3";
+    private static final String FILE_2_4 = "file_name_2_4";
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: This is quite a lot of cases. Are they all needed?

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+    private static final String FILE_BULKSCAN_1 = "file_name_bulkscan_1";
+
+    @BeforeEach
+    void setUp() {
+        dbHelper.deleteAll();
+    }
+
+    @Test
+    @SuppressWarnings("checkstyle:variabledeclarationusagedistance")
+    void should_find_within_date_range() {
```
Comment (lgonczar) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: This test tests quite a lot at once. It would be worth splitting it into multiple tests, each focused on a different thing. For example:
- that the retrieved data maps correctly to fields
- that the search doesn't return records from outside given date range
- that the search doesn't return results for a different container
- that the search doesn't return results for a different file
- event data comes from the last event

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java (hunk 6 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java]: well. event i always present. whereas envelope is created at the certain condition. since there's no link, should just trust on events first

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java (hunk 7 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: out of the box benefit: automatic escape and quoting

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java (hunk 6 above)
Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryMapper.java]: I suggest bullet proof approach

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+    public EnvelopeSummaryRepository(
+        NamedParameterJdbcTemplate jdbcTemplate,
+        EnvelopeSummaryMapper mapper
+    ) {
+        this.jdbcTemplate = jdbcTemplate;
+        this.mapper = mapper;
+    }
+
+    public List<EnvelopeSummary> find(Instant from, Instant to) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepository.java]: clearer name or javadoc would be nice? which date are `from` and `to` used for?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EnvelopeSummary.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+
+    private final String container;
+    private final String fileName;
+    private final Instant fileCreatedAt;
+    private final Instant dispatchedAt;
+    private final Status status;
+    private final boolean isDeleted;
+    private final Event lastEvent;
+    private final String lastEventNotes;
+    private final Instant lastEventCreatedAt;
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/data/model/EnvelopeSummary.java]: shouldn't those be public?

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java (hunk 14 above)
Comment (rkondratowicz) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: agree

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+                    null,
+                    null,
+                    null
+                )
+            );
+    }
+
+    @Test
+    @SuppressWarnings("checkstyle:variabledeclarationusagedistance")
+    void should_skip_bulkscan_repository() {
```
Comment (rkondratowicz) [inline src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: what's tested here?

Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: Envelopes from bulkscan repository should not be included in reports.

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java (hunk 14 above)
Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: @rkondratowicz @lgonczar 
There are 3 test methods:
- should_find_within_date_range - checks that files before and after given time range are filtered out, also (as well as other methods) it checks that all fields in EventSummary have correct values including container and file name and that event related fields have their values from the last event
- should_skip_bulkscan_repository - checks that the file from 'bulkscan' container is not included in the result
- should_handle_missing_events - checks that in case of no events for a file the result is still successfully fetched; though we expect that every file should have at least 1 event handling this situation (no events for a file) makes the code base bullet proof

I hope this explains details of the test class

File: src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java (hunk 17 above)
Comment (rkondratowicz) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: you mean 'container'?

Comment (apereverzin) [reply src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/data/EnvelopeSummaryRepositoryTest.java]: @rkondratowicz Yes, renamed the method

**Comment by rkondratowicz** [comment]
Given https://github.com/hmcts/blob-router-service/pull/298 exists, is this PR still relevant?
If not, can it be closed?

**Comment by apereverzin** [comment]
PR-298
